src/to-car.ts

```diff
@@ -0,0 +1,101 @@
+import fs from 'fs'
+import os from 'os'
+import path from 'path'
+import process from 'process'
+import { Readable, Writable } from 'stream'
+
+import { CarWriter } from '@ipld/car'
+import { importer } from 'ipfs-unixfs-importer'
+import { UnixFSEntry } from 'ipfs-unixfs-exporter'
+
+import { CID } from 'multiformats'
+import all from 'it-all'
+import pipe from 'it-pipe'
+
+import { sha256 } from 'multiformats/hashes/sha2'
+
+import normalizeAddInput from 'ipfs-core-utils/src/files/normalise-input/index'
+
+import globSource from 'ipfs-utils/src/files/glob-source'
+import { WriterChannel } from '@ipld/car/lib/writer'
+
+export async function packFileToCarFs ({ input, output }: { input: string | Iterable<string> | AsyncIterable<string>, output?: string }) {
+  const location = output || `${os.tmpdir()}/${(parseInt(String(Math.random() * 1e9))).toString(36) + Date.now()}`
+  const writable = fs.createWriteStream(location)
+
+  const root = await packFileToCar({input, writable})
+
+  // Move to work dir
+  if (!output) {
+    const inputName = typeof input === 'string' ? path.parse(path.basename(input)).name : root.toString()
+    await fs.promises.rename(location, `${process.cwd()}/${inputName}.car`)
+  }
+}
+
+export async function packFileToCar ({ input, writable }: { input: string | Iterable<string> | AsyncIterable<string>, writable: Writable}) {
+  let writerChannel: WriterChannel | undefined
+  let root: CID | undefined
+
+  const blockStore: Array<{ cid: CID, bytes: Uint8Array }> = []
+
+  const blockApi = {
+    put: ({ cid, bytes }: { cid: CID, bytes: Uint8Array}) => {
+      root = cid
+      blockStore.push({ cid, bytes })
+
+      return Promise.resolve({ cid, bytes })
+    },
+    get: (cid: CID) => {
+      return Promise.reject(new Error('should not get blocks'))
+    }
+  }
+
+  // Consume the source
+  await all(pipe(
+    normalizeAddInput(globSource(input, {
+      recursive: true
+    })),
+    (source: any) => importer(source, blockApi, {
+      cidVersion: 1,
+      chunker: 'fixed',
+      maxChunkSize: 262144,
+      hasher: sha256,
+      rawLeaves: true,
+      wrapWithDirectory: false // TODO: Set to true when not directory to keep names?
+    }), // TODO: recheck defaults
+  ))
+
+  if (!blockStore.length || !root) {
+    throw new Error('Not valid file to pack')
+  }
+
+  writerChannel = CarWriter.create([blockStore[blockStore.length - 1].cid])
+  Readable.from(writerChannel.out).pipe(writable)
+  
+  for (let i = 0; i < blockStore.length; i++) {
+    await writerChannel.writer.put(blockStore[i])
+  }
+  await writerChannel.writer.close()
+
+  return root
+}
+
+// UnixFs to Car function
+export async function toCar ({ files, writable }: { files: UnixFSEntry[], writable: Writable}) {
```

Comment:
Let's rethink this one. I think we want to provide an API closer to what `ipfs.add` and `ipfs.addAll` that take an iterable of `FileObject` or `FileContent`  which allows the caller to pass in objects that define the attributes of the files for the bytes they are passing in.

```
{
  // The path you want the file to be accessible at from the root CID _after_ it has been added
  path?: string
  // The contents of the file (see below for definition)
  content?: FileContent
  // File mode to store the entry with (see https://en.wikipedia.org/wiki/File_system_permissions#Numeric_notation)
  mode?: number | string
  // The modification time of the entry (see below for definition)
  mtime?: UnixTime
}
```

https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/FILES.md#ipfsadddata-options

Reply:
happy to discuss and reconsider. What are the advantages you seen on having this API closer to `ipfs.add`? The [`UnixFsEntry`](https://github.com/ipfs/js-ipfs-unixfs/blob/js-dag-pb/packages/ipfs-unixfs-exporter/src/types.d.ts#L49) is really similar. In my opinion, the main advantage of using this type is that we can do:

```js
const files = await all(fromCar(carReader))
await toCar({
      files,
      writable: fs.createWriteStream(`${__dirname}/tmp/raw.car`)
    })
```

Having `fromCar` returning something that is not consumable by `toCar` does not feel like the best approach to me. What do you think?

Reply:
```
// File mode to store the entry with (see https://en.wikipedia.org/wiki/File_system_permissions#Numeric_notation)
  mode?: number | string
  // The modification time of the entry (see below for definition)
  mtime?: UnixTime
```

Do you know where these 2 metadata items end up? They do not seem part of the unixfs side

Reply:
ah ok. We can keep `fromCar` as the opposite of `toCar`. I'm more thinking about how to use this lib from a browser, which can wait for a subsequent PR.